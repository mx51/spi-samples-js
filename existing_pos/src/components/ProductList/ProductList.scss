@import '../ColorPalette';


.product-display {
  display: inline-block;
  border: 1px solid $primaryGrey;
  margin: 5px;
  text-align: center;
  padding: 0;
}
.product-display img {
  width: 100%;
}
.product-category {
  font-size: 1.2rem;
  margin-top: 2rem;
  font-family: 'Muli-Regular', 'Muli';
  padding: 0 0 1rem;
}
.product-display:focus {
  border-color: $primaryRed;
  border: 2px solid $primaryRed;
  outline: none;
}
.product-display:hover{
  border: 1px solid $primaryRed;
}
