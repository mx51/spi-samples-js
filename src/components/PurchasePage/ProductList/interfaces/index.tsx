export interface IProduct {
  name: string;
  price: string;
  image: React.ReactElement;
  backgroundColor: string;
}
