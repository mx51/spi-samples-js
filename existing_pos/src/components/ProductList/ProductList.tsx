import React from 'react';
import { Category } from '../types';
import './ProductList.scss';

function ProductList(props: { category: Category; onProductClick: Function }) {
  const { category, onProductClick } = props;
  return (
    <div>
      <h3 className="product-category">{category.categoryName}</h3>
      <hr />
      {category.list.map((p) => (
        <span key={`${category.categoryName}-${p.id}`}>
          <button
            id={`btnProductItem${p.id}`}
            type="button"
            className="product-display"
            onClick={() => onProductClick(p.id)}
          >
            <img src={p.image} alt={p.name} />
            <p> {p.name} </p>
          </button>
        </span>
      ))}
    </div>
  );
}

export default ProductList;
