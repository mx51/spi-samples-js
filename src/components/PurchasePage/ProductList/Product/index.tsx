import React from 'react';
import { Box, Card, CardActionArea, CardContent, CardMedia, Typography } from '@material-ui/core';
import useStyles from './index.styles';
import { ProductProps } from './interfaces';

export default function Product({ product }: ProductProps): React.ReactElement {
  const classes = useStyles();
  return (
    <Box>
      <Card className={classes.cardStyle}>
        <CardActionArea>
          <CardMedia className={classes.productText} title={product.name}>
            <Box className={classes.productImage} display="flex" justifyContent="center">
              <img src={product.image} alt={product.name} />
            </Box>
          </CardMedia>
          <CardContent>
            <Typography className={classes.productName} variant="h5" component="h2">
              {product.name}
            </Typography>
            <Typography variant="body2" color="textSecondary" component="p">
              {Intl.NumberFormat('en-AU', { style: 'currency', currency: 'AUD' }).format(product.price)}
            </Typography>
          </CardContent>
        </CardActionArea>
      </Card>
    </Box>
  );
}
