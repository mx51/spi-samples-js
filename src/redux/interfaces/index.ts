import { IProductState } from '../reducers/ProductSlice/interfaces';
import { ITerminalState } from '../reducers/TerminalSlice/interfaces';

export interface ISamplePosState {
  terminals: ITerminalState;
  products: IProductState;
}
