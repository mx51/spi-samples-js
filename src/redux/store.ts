import { AnyAction, Reducer, combineReducers, configureStore } from '@reduxjs/toolkit';
import { createMigrate, persistReducer, persistStore } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import hardSet from 'redux-persist/es/stateReconciler/hardSet';
import commonReducer from './reducers/CommonSlice/commonSlice';
import pairFormReducer from './reducers/PairFormSlice/pairFormSlice';
import { preAuthReducer } from './reducers/PreAuth/preAuthSlice';
import productSlice from './reducers/ProductSlice/productSlice';
import selectedTerminalReducer from './reducers/SelectedTerminalSlice/selectedTerminalSlice';
import terminalReducer from './reducers/TerminalSlice/terminalsSlice';
import { IPreAuthState } from './reducers/PreAuth/interfaces';
import { PayAtTableState, payAtTableReducer } from './reducers/PayAtTableSlice/payAtTableSlice';
import { ITerminalState } from './reducers/TerminalSlice/interfaces';

const terminalReducerMigration: Any = {
  0: (state: ITerminalState) => {
    Object.keys(state).forEach((key) => {
      if (state[key] && state[key].deviceAddress) {
        state[key].deviceAddress = state[key].deviceAddress.replace(/^wss?:\/\//, '');
      }
      if (state[key] && state[key].secrets) {
        state[key].secrets = {
          encKey: (state[key].secrets as Any).EncKey,
          hmacKey: (state[key].secrets as Any).HmacKey,
        };
      }
    });
    return state;
  },
};

const rootReducer = combineReducers({
  common: commonReducer,
  pairForm: pairFormReducer,
  products: productSlice,
  terminals: persistReducer<ITerminalState, AnyAction>(
    {
      key: 'terminals',
      version: 0,
      storage,
      stateReconciler: hardSet,
      migrate: createMigrate(terminalReducerMigration, { debug: false }),
    },
    terminalReducer
  ) as unknown as Reducer<ITerminalState>,
  selectedTerminal: selectedTerminalReducer,
  preAuth: persistReducer<IPreAuthState, AnyAction>(
    {
      key: 'root',
      storage,
      stateReconciler: hardSet,
    },
    preAuthReducer
  ) as unknown as Reducer<IPreAuthState>,
  payAtTable: persistReducer<PayAtTableState, AnyAction>(
    {
      key: 'payAtTable',
      storage,
      stateReconciler: hardSet,
    },
    payAtTableReducer
  ) as unknown as Reducer<PayAtTableState>,
});

export const store = configureStore({
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
  reducer: rootReducer,
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = typeof store;
export type AppDispatch = typeof store.dispatch;
