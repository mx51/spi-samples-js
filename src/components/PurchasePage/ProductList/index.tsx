import React from 'react';
import { Box } from '@material-ui/core';
import products from '../../../definitions/constants/productList';
import Product from '../Product';

function ProductList(): React.ReactElement {
  return (
    <Box display="flex" flexDirection="coloum" flexWrap="wrap" justifyContent="space-around">
      {products.map((product) => (
        <Product key={product.name} product={product} />
      ))}
    </Box>
  );
}
export default ProductList;
