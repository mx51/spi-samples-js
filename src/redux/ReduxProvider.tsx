import React from 'react';
import { Provider } from 'react-redux';
// store
import { store } from './store';

interface ReduxProviderInterface {
  children: React.ReactNode;
  reduxStore: typeof store;
}

function ReduxProvider({ children, reduxStore }: ReduxProviderInterface): React.ReactElement {
  return <Provider store={reduxStore}>{children}</Provider>;
}

export default ReduxProvider;
