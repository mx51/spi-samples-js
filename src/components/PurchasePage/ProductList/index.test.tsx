import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import * as redux from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import '@testing-library/jest-dom/extend-expect';
import ProductList from '.';
import ReduxProvider from '../../../redux/ReduxProvider';
import { store } from '../../../redux/store';

describe('Test <ProductList />', () => {
  test('snapshot for ProductList component', () => {
    const { container } = render(
      <ReduxProvider reduxStore={store}>
        <Router>
          <ProductList />
        </Router>
      </ReduxProvider>
    );
    expect(container).toMatchSnapshot();
  });

  test('addProduct when clicking product', () => {
    const spy = jest.spyOn(redux, 'useDispatch');
    const dispatch = jest.fn();
    spy.mockReturnValue(dispatch);

    const { getByText } = render(
      <ReduxProvider reduxStore={store}>
        <Router>
          <ProductList />
        </Router>
      </ReduxProvider>
    );

    fireEvent.click(getByText(/Mocha/i));
    expect(dispatch.mock.calls.length).toBe(2);
    expect(dispatch.mock.calls[0][0]).toEqual({
      payload: 0,
      type: 'product/addKeypadAmount',
    });
    expect(dispatch.mock.calls[1][0]).toEqual({
      payload: { product: { id: 1, image: 'images/product/mocha.png', name: 'Mocha', price: 420 } },
      type: 'product/addProduct',
    });
  });
});
