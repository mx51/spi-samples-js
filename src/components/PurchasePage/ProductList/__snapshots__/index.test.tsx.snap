// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ProductList /> snapshot for ProductList component 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <div
      class="MuiBox-root MuiBox-root-6"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Mocha"
          >
            <div
              class="MuiBox-root MuiBox-root-7 makeStyles-productImage-5"
            >
              <img
                alt="Mocha"
                src="images/product/mocha.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Mocha
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $4.20
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-8"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Honey Latte"
          >
            <div
              class="MuiBox-root MuiBox-root-9 makeStyles-productImage-5"
            >
              <img
                alt="Honey Latte"
                src="images/product/honeyLatte.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Honey Latte
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $4.80
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-10"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Flat White"
          >
            <div
              class="MuiBox-root MuiBox-root-11 makeStyles-productImage-5"
            >
              <img
                alt="Flat White"
                src="images/product/flatWhite.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Flat White
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $3.80
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-12"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Chai Latte"
          >
            <div
              class="MuiBox-root MuiBox-root-13 makeStyles-productImage-5"
            >
              <img
                alt="Chai Latte"
                src="images/product/chaiLatte.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Chai Latte
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $4.50
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-14"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Rose Tea Latte"
          >
            <div
              class="MuiBox-root MuiBox-root-15 makeStyles-productImage-5"
            >
              <img
                alt="Rose Tea Latte"
                src="images/product/roseTeaLatte.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Rose Tea Latte
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $4.50
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-16"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Cappuccino"
          >
            <div
              class="MuiBox-root MuiBox-root-17 makeStyles-productImage-5"
            >
              <img
                alt="Cappuccino"
                src="images/product/cap.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Cappuccino
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $3.80
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-18"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Latte"
          >
            <div
              class="MuiBox-root MuiBox-root-19 makeStyles-productImage-5"
            >
              <img
                alt="Latte"
                src="images/product/latte.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Latte
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $3.80
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-20"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-cardStyle-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-productText-3"
            title="Production Test Item"
          >
            <div
              class="MuiBox-root MuiBox-root-21 makeStyles-productImage-5"
            >
              <img
                alt="Production Test Item"
                src="images/product/testItem.png"
              />
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-productName-4 MuiTypography-h5"
            >
              Production Test Item
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $0.01
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
